controllers/cloudstackmachine_controller.go

```diff
@@ -153,41 +156,55 @@ func (r *CloudStackMachineReconciler) reconcile(
 	csCluster *infrav1.CloudStackCluster) (ctrl.Result, error) {
 
 	log.V(1).Info("reconcile CloudStackMachine")
-
 	// Make sure bootstrap data is available in CAPI machine.
 	if capiMachine.Spec.Bootstrap.DataSecretName == nil {
 		log.Info("Bootstrap DataSecretName not yet available.")
 		return ctrl.Result{}, nil
 	}
 	log.Info("Got Bootstrap DataSecretName.")
 
+	// Set ZoneID on csMachine.
+	if util.IsControlPlaneMachine(capiMachine) { // Use failure domain zone.
+		csMachine.Status.ZoneID = *capiMachine.Spec.FailureDomain
+	} else { // Random zone.
+		zones := make([]string, len(csCluster.Status.Zones))
+		zidx := 0
+		for zoneID := range csCluster.Status.Zones {
+			zones[zidx] = zoneID
+			zidx++
+		}
+		rand.Seed(time.Now().Unix())
```

Comment:
I believe re-seeding the rand every invocation thwarts its randomness.  And is a cost that should only be borne once.  I think this should be called in main.go.

Reply:
I think it does the opposite since the time it's seeded is random too, but I can move it to main if you'd like?

Reply:
Yes, please.  I'd rather put my trust in the randomness of the algorithm.

Reply:
I disagree that it thwarts the randomness in any meaningful way, at least if the algorithm is implemented correctly.  But it's also completely unnecessary to seed it more than once.  I agree with moving it to main.

Reply:
Actually, it can thwart the randomness if you seed it twice in a short enough time that you use the same seed.